File: Day05/part02.py
import os

file = os.path.join(os.path.dirname(__file__), "blob.txt")


def file_reader(file: str) -> str:
    with open(file, "r") as f:
        return f.read()


def solver(s: str) -> int:
    maximum = 0
    seats = []
    for options in s.splitlines():
        options = (
            options.strip()
            .replace("F", "0")
            .replace("B", "1")
            .replace("L", "0")
            .replace("R", "1")
        )
        seats.append(int(options, 2))
        maximum = max(maximum, int(options, 2))

    for i in range(min(seats), maximum):
        if i not in seats:
            return i
    return 0


if __name__ == "__main__":
    print(solver(file_reader(file)))

File: Day05/part02.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func fileReader(file string) []string {

	f, err := os.Open(file)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()
	var ret []string
	scanner := bufio.NewScanner(f)

	for scanner.Scan() {
		ret = append(ret, scanner.Text())
	}

	return ret
}

func minValue(lst []int64) int64 {
	var minimum int64 = 9999999

	for _, l := range lst {
		if l <= minimum {
			minimum = l
		}
	}
	return minimum

}

func maxValue(lst []int64) int64 {
	var maximum int64 = 0

	for _, l := range lst {
		if maximum <= l {
			maximum = l
		}
	}

	return maximum

}

func seatNotFound(a int64, list []int64) bool {
	for _, b := range list {
		if b == a {
			return false
		}
	}
	return true
}

func seatFinder(s []string) int64 {
	var seats []int64

	for _, s := range s {
		s = strings.Replace(strings.Replace(
			strings.Replace(
				strings.Replace(
					s, "F", "0", -1),
				"B", "1", -1),
			"L", "0", -1),
			"R", "1", -1)
		intS, err := strconv.ParseInt(s, 2, 64)

		if err != nil {
			log.Fatal(err)
		}

		seats = append(seats, intS)
	}

	for i := minValue(seats); i <= maxValue(seats); i++ {
		if seatNotFound(i, seats) {
			return i
		}
	}
	return 0
}

func main() {
	fmt.Println(seatFinder(fileReader("./blob.txt")))
}
